### PR TITLE
Full package review for {bpmodels} CRAN release

### PR DESCRIPTION
**Pull request by jamesmbaazam**
This is a full package review in preparation for submitting {bpmodels} to CRAN. It would be helpful to provide comments that help resolve the checklist provided here #80.

# Modalities

## How to become a full reviewer

Self-assignment under the "Reviewers" tab on the right side of this window.

## Required number of reviewers

Maximum of 2. PS: Don't self-assign when this is reached.

### Full reviewers
1. @joshwlambert 
2. @pratikunterwegs 

## Deadline

Not currently set.

## What next?

- This PR will be closed when all relevant issues have been resolved. 
- Closing this PR will not close any issues raised but will serve as a reference for raising issues to be resolved in subsequent pull requests.

### CODE REVIEW

File: DESCRIPTION [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    ggplot2,
+    knitr,
+    lubridate,
+    rmarkdown,
+    testthat,
+    truncdist
+VignetteBuilder:
+    knitr
+Remotes:
+    github::epiverse-trace/epiparameter
```
Comment (Bisaloo) [inline DESCRIPTION:39]: I didn't go for a full review just yet but I want to flag early that you cannot use `Remotes` on CRAN. There are alternatives but they have their own issues. Your best bet is to:
- either wait that epiparameter is on CRAN
- or drop epiparameter from `Suggests`

Comment (jamesmbaazam) [reply DESCRIPTION:39]: Thanks for flagging @Bisaloo. Do you intend to do a full review or is the second slot open?

Comment (Bisaloo) [reply DESCRIPTION:39]: I'm happy to let anyone else do it so it's still open. Let me know if you don't find anyone else.

Comment (jamesmbaazam) [reply DESCRIPTION:39]: I will. Thanks.

Comment (pratikunterwegs) [reply DESCRIPTION:39]: I can volunteer, will self assign now

Comment (jamesmbaazam) [reply DESCRIPTION:39]: > * or drop epiparameter from `Suggests`

I think this is the more practical solution. `epiparameter` is only used in a few places in the vignette and should be straightforward to change here ae24211d5708d4c14501feb2cf88b96703530338. I will edit the vignette to remove `epiparameter` and return to this commit when `epiparameter` is put on CRAN.

**Comment by joshwlambert** [comment]
Overall package is in good shape and functionally is definitely ready for release. Below are comments for some of the functions and other aspects of the package. 

I have avoided requesting any breaking changes, with the aim to provide a stable, backwards-compatible release.

### General Package comment

- gitattributes empty file in package root, can this be deleted? It is not present in other Epiverse-TRACE packages
- lint_changed_files.yaml can be removed and switched to using the organisation-wide lint workflow.
- updated R-CMD-check.yaml workflow to match {packagetemplate}
- update render-readme.yaml workflow to match {packagetemplate}
- update test-coverage.yaml workflow to match {packagetemplate}
- Everything else in .github looks good
- Not a big issue, but I would move the modelling section above the literature section in \_pkgdown.yml  as it is a more important vignette.
- update .lintr to match {packagetemplate}
- I would recommend removing the citation.cff file, for the following reasons:
	- it requires you to manually update both the DESCRIPTION and the citation file
	- the citation file is currently out of date
	- other Epiverse-TRACE packages do not have one and as a result the citation style differs between {bpmodels} and the other packages. (If you have a strong case for the file I would recommend having it become standardised across all Epiverse packages)
	- it would remove the need for the inst/ dir
- data and data-raw looks fine
- Could add the word "simulating" somewhere in the package title in DESCRIPTION as currently only covers the inference part of the package
- Suggest moving the maintainer to the first author in the author list in the DESCRIPTION
- There are some CRAN rules around the formatting of citations in the package description, would recommend checking these before submitting.
- Add spell checking to package
- I recommend future releases use either consistent subheaders in the NEWS.md file (e.g.{EpiNow2} <https://github.com/epiforecasts/EpiNow2/blob/main/NEWS.md>) or not using subheaders (e.g. finalsize <https://github.com/epiverse-trace/finalsize/blob/main/NEWS.md>). This makes it easier to scan, especially when there are a large number of releases

### Readme and vignettes

- If you change the package description to include "simulating", update readme header to match
- Currently there are two `knitr::opts_chunk$set()` calls in the README, these can probably be collapsed into a single call
- I would use explicit argument matching in the README as it will likely be read by people new to the package and avoid position matching which can lead to issues for users.
- Consider changing "poisson offspring with mean, lambda = 0.5" to "poisson offspring with mean 0.5 (`lambda = 0.5`)"
- When calling `chain_sim_susc()` in the README each argument could go on it's own line
- Is there a reason for having two files for references, might be easier to merge them into a single file containing all required refs
- Reproduction number does not render correctly in get started vignette
- It might be worth mentioning somewhere in the get started vignette that the simulation is stochastic and therefore setting the seed is required for consistent output. This could be in the text or a small comment above the `set.seed()`
- In the get started vignette, is a vector of four chains sizes a good example to use? I don't know, but I would imagine the power to differentiate between generating models is low when the sample size is four (assuming that the sample size is the number of transmission chains and not the total size of all chains)
- Recommend rewording "The second argument, `offspring`, specifies the offspring distribution. This is given as a function used to generate random offspring." as this argument requires a character, not a closure (function). This is also checked on L149 in likelihood.R
- The 
```r
?chain_ll
```
could go inline as it is not evaluated and is a short chunk of code.
- The code chunk
```r

chain_sizes <- c(1, 1, 4, 7) # example of observed chain sizes
chain_ll(chain_sizes, "binom", "size", size = 1, prob = 0.5,
         nsim_offspring = 100)
```
returns `-Inf` is this correct? Could it be because you're using the binomial instead of the negative binomial?
In the getting started vignette setting the seed once is enough to consistently produce the vignette unless you change the number of times the PRNG is sampled.
- I like the literature vignette, is the plan to port this over to {epichains}?
- In the COVID vignette you could simplify the `mapply()` call, see
``` r
library("bpmodels")
days_since_index = c(0, 2, 3, 4, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15)
data("covid19_sa")
covid19_sa
#>          date cases
#> 1  2020-03-05     1
#> 2  2020-03-07     1
#> 3  2020-03-08     1
#> 4  2020-03-09     4
#> 5  2020-03-11     6
#> 6  2020-03-12     3
#> 7  2020-03-13     8
#> 8  2020-03-14    14
#> 9  2020-03-15    13
#> 10 2020-03-16    11
#> 11 2020-03-17    23
#> 12 2020-03-18    31
#> 13 2020-03-19    34
#> 14 2020-03-20    52
start_times <- unlist(mapply(
  function(x, y) rep(x, times = ifelse(y == 0, 1, y)),
  days_since_index,
  covid19_sa$cases
))
start_times2 <- rep(days_since_index, covid19_sa$cases)
identical(start_times, start_times2)
#> [1] TRUE
```

<sup>Created on 2023-06-23 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>
However, your use of `ifelse()` makes me think you don't want any zeros in the start times, but there is one at the start. I'm perhaps misinterpreting what this code should be doing, but please check and simplify if possible
- In the COVID vignette, I recommend using `argument` and `function()` syntax to differentiate. Currently `serial_interval` is missing parentheses.
- The `return()` and `si <- ` call could be removed from:
```r
#' serial interval function
serial_interval <- function(sample_size) {
  si <- rlnorm(sample_size, meanlog = log_mean, sdlog = log_sd)
  return(si)
}
```
to make it consistent with the get started vignette.
- I would reword "The negative binomial distribution is commonly used in epidemiology to account for individual variation in transmissibility, also known as superspreading". As small amounts of variation (e.g. poisson or geometric) are not considered superspreading.
- I think it would be good to justify why `chain_threshold` is set to 1000.
- It seems that in figure 1 of the COVID vignette many of the grey lines are above the red median line. I would have expected the red line to sit in the middle of each stochastic simulation?
- Would remove "(median)" from figure 1 y-axis

### Functions

- I think a sensible default for `nsim_obs` in `chain_ll()` would be useful. Without detailed domain knowledge for these things it can be hard to judge a sensible magnitude and changing will scale the run time. Should it be 100, 1000, 1,000,000? By giving a sensible default it will help beginners and doesn't prevent experienced users from modifying. It is also backwards compatible.
- It might be good to have a control on simulating negative serial intervals. For now you could limit the functionality to work with strictly positive serial interval. E.g.
``` r
library("bpmodels")
set.seed(1)
serial_interval <- function(n) {
  rnorm(n, mean = 0.5, sd = 2)
}
serial_interval(10)
#>  [1] -0.7529076  0.8672866 -1.1712572  3.6905616  1.1590155 -1.1409368
#>  [7]  1.4748581  1.9766494  1.6515627 -0.1107768
chains_df <- chain_sim(
  n = 5, offspring = "pois", lambda = 0.5, stat = "length",
  infinite = 100, serial = serial_interval, tree = TRUE
)
chains_df
#>   n id ancestor generation        time
#> 1 1  1       NA          1  0.00000000
#> 2 2  1       NA          1  0.00000000
#> 3 3  1       NA          1  0.00000000
#> 4 4  1       NA          1  0.00000000
#> 5 5  1       NA          1  0.00000000
#> 6 1  2        1          2 -0.07892315
#> 7 1  3        1          2 -0.09843024
#> 8 3  2        1          2 -0.32302167
```

<sup>Created on 2023-06-23 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

- Should the `dborel()` have input checking to make sure the parameter `mu` is between 0 and 1? If so this may also apply to `rborel()`.
- Function documentation and data documentation look good (bpmodels-package.R and data.R)
- For the internal functions in likelihood.R you could also use the `@noRd` tag to prevent the documentation files from being produced. I assume you are only using roxygen doc in these functions for the developers rather than users.
- Can replace `\code`{} & `\link{}` roxygen tags with markdown syntax
- I find using `numeric(1)` or `character(1)` improves readability for `vapply()` but it's personal preference
- `complementary_logprob()` return documentation is `value`. Could give this a better description?
- `find_function_name()` is internal not used in the package. Not suggesting it's removal, but would be good to evaluate why it's in the package.
- Given that the aim is for a stable release I did not look into whether the simulation functions could be optimised.
- In `chain_sim_susc()` I would recommend changing the `@param pop the population` to `@param pop an integer of the population size`.

**Comment by joshwlambert** [comment]
@jamesmbaazam I'm was happy to review promptly given our discussions around the release, however, I think the deadline set in the first comment on this PR is too short. It gives people less than a week to find time to conduct a full package review. I would recommend extending this deadline to give others a chance to comment, especially as this PR is not a urgent bottleneck for other projects.

**Comment by jamesmbaazam** [comment]
> I would recommend removing the citation.cff file, for the following reasons:
> * it requires you to manually update both the DESCRIPTION and the citation file
> * the citation file is currently out of date
> * other Epiverse-TRACE packages do not have one and as a result the citation style differs between {bpmodels} and the other packages. (If you have a strong case for the file I would recommend having it become standardised across all Epiverse packages)
> * it would remove the need for the inst/ dir

I would suggest adding it to {packagetemplate} because it adds a button on the right of the repo that makes it easy to copy the citation information (see attached). If we're advocating for packages to be cited, we should make it as easy as possible. I would like to hear @Bisaloo's thoughts on this.
Screenshot 2023-07-12 at 18 04 08

**Comment by jamesmbaazam** [comment]
> * lint_changed_files.yaml can be removed and switched to using the organisation-wide lint workflow.

Thanks. How do I switch to the org-wide workflow?

**Comment by Bisaloo** [comment]
> Thanks. How do I switch to the org-wide workflow?

You're already using it. You can see it in the following PR by the fact that you have two `lint-changed-files`. One is the org-wide one, and the other is from this repo.

https://github.com/epiverse-trace/bpmodels/pull/89

**Comment by pratikunterwegs** [comment]
Hi @jamesmbaazam would you like me to add my review, or do you want me to wait some more until you've handled Josh's review?

**Comment by jamesmbaazam** [comment]
Hi @pratikunterwegs, I think I'll work on Josh's for now. Thanks.

**Comment by jamesmbaazam** [comment]
> * There are some CRAN rules around the formatting of citations in the package description, would recommend checking these before submitting.

I checked the current format against what was suggested in the [extrachecks](https://github.com/DavisVaughan/extrachecks#you-get-asked-if-there-are-any-references-describing-the-methods-in-your-package) list and it appears to conform.

**Comment by jamesmbaazam** [comment]
> * I would use explicit argument matching in the README as it will likely be read by people new to the package and avoid position matching which can lead to issues for users.

Good catch. Totally agree!

**Comment by jamesmbaazam** [comment]
> * Consider changing "poisson offspring with mean, lambda = 0.5" to "poisson offspring with mean 0.5 (`lambda = 0.5`)"

Could you explain how they're different? `lambda` as used here is consistent with how R and the broader stats community know it, so I don't think it's confusing at all.

**Comment by jamesmbaazam** [comment]
> * Is there a reason for having two files for references, might be easier to merge them into a single file containing all required refs

Yes, there's a reason. `branching_process_literature.json` is specifically a database of modelling exercises that used branching processes. It is printed as a bibliography in the `branching_process_literature.Rmd` vignette and keeping it separate is convenient for that. It is also meant to serve as a database for others to contribute to, for example, this [PR](https://github.com/epiverse-trace/bpmodels/pull/78). On the other hand, `vignettes/references.json` contains other literature that may fall outside of what is contained in `branching_process_literature.json`.

**Comment by jamesmbaazam** [comment]
> * In the get started vignette, is a vector of four chains sizes a good example to use? I don't know, but I would imagine the power to differentiate between generating models is low when the sample size is four (assuming that the sample size is the number of transmission chains and not the total size of all chains)

I'm not sure I understand.

**Comment by jamesmbaazam** [comment]
> * I like the literature vignette, is the plan to port this over to {epichains}?

Thanks. Yes, that's the plan. The goal is to make it easier for people interested in the subject to find resources right here in the package.

**Comment by jamesmbaazam** [comment]
> However, your use of `ifelse()` makes me think you don't want any zeros in the start times, but there is one at the start. I'm perhaps misinterpreting what this code should be doing, but please check and simplify if possible

I think the goal is to create a start time for dates with 0 cases. For a large dataset, the utility of this setup becomes more apparent.

**Comment by jamesmbaazam** [comment]
> * It seems that in figure 1 of the COVID vignette many of the grey lines are above the red median line. I would have expected the red line to sit in the middle of each stochastic simulation?

That's a good point. I'll crosscheck that.

**Comment by jamesmbaazam** [comment]
> * It might be good to have a control on simulating negative serial intervals. For now you could limit the functionality to work with strictly positive serial interval.

I am not sure that this is a necessary feature as we do not want to restrict the user on what types of serial intervals to simulate. There's a case for users wanting to simulate negative serial intervals to pursue questions about the infector being infectious before developing symptoms (See the "serial interval" section of [this paper](https://academic.oup.com/aje/article/180/9/865/2739204) that for example mentions that observed serial intervals can be negative, and [this one](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7239082/#:~:text=Negative%20serial%20intervals%20(left%20of,asymptomatic%20or%20mildly%20symptomatic%20cases) that observed negative serial intervals, and other examples in the literature. 

@sbfnk Thoughts on this requested feature?

**Comment by jamesmbaazam** [comment]
> * The code chunk
> 
> ```r
> chain_sizes <- c(1, 1, 4, 7) # example of observed chain sizes
> chain_ll(chain_sizes, "binom", "size", size = 1, prob = 0.5,
>          nsim_offspring = 100)
> ```
> 
> returns `-Inf` is this correct? 
> Could it be because you're using the binomial instead of the negative binomial?

There is an analytical solution for the negative binomial size distribution. This code chunk demos the simulated binomial size result.

**Comment by sbfnk** [comment]
> > * In the get started vignette, is a vector of four chains sizes a good example to use? I don't know, but I would imagine the power to differentiate between generating models is low when the sample size is four (assuming that the sample size is the number of transmission chains and not the total size of all chains)

It's a good point - 

> > * It might be good to have a control on simulating negative serial intervals. For now you could limit the functionality to work with strictly positive serial interval.
> 
> I am not sure that this is a necessary feature as we do not want to restrict the user on what types of serial intervals to simulate. There's a case for users wanting to simulate negative serial intervals to pursue questions about the infector being infectious before developing symptoms (See the "serial interval" section of [this paper](https://academic.oup.com/aje/article/180/9/865/2739204) that for example mentions that observed serial intervals can be negative, and [this one](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7239082/#:~:text=Negative%20serial%20intervals%20(left%20of,asymptomatic%20or%20mildly%20symptomatic%20cases) that observed negative serial intervals, and other examples in the literature.
> 
> @sbfnk Thoughts on this requested feature?

I agree - supporting negative SIs is a feature not a bug.

**Comment by jamesmbaazam** [comment]
All issues raised through this full package review have been addressed and logged as GitHub issues, so I am closing this PR now.
